dohproxy/server_protocol.py

```diff
@@ -42,10 +43,34 @@ def __init__(self, upstream_resolver, upstream_port, logger=None):
         self.logger = logger
         self.transport = None
 
-    async def query(self, dnsq, clientip, timeout=DEFAULT_TIMEOUT):
-        dnsr = await self.query_udp(dnsq, clientip, timeout=timeout)
+    async def query(self, dnsq, clientip, timeout=DEFAULT_TIMEOUT,
+                    ecs=False):
+        original_ecs_option = None
+        for option in dnsq.options:
+            if isinstance(option, dns.edns.ECSOption):
+                original_ecs_option = option
+                break
+
+        # (Potentially) modified copy of dnsq
+        dnsq_mod = dns.message.from_wire(dnsq.to_wire())
```

Comment:
are you duplicating the `dnsq` in order to be able to access `dnsq.edns` at a later stage? Maybe it would suffice to save the original edns value along original_ecs_option?

Reply:
Minorly yes, but more importantly since dnsq is a mutable object and is used farther up the chain, to make responses by the proxy if there's an error.  I could try to restore dnsq to its previous unaltered state at the end of query(), but there's the possibility it doesn't look exactly the same as when it went in.  Making a copy of dnsq, potentially altering it, and sending that on the wire is easier than trying to modify and then restore it.

Reply:
sgtm